### ISSUE TITLE
Include covariant interfaces for Map, Set and Array

### ISSUE THREAD

**Issue by jesseschalken** [facebook/flow#2333]
Code is often written in terms of arrays with the expectation that if `B` fits `A` then `Array<B>` fits `Array<A>`. This is not true because `Array` is (correctly) invariant in its type parameter due to methods such as `push()`.

``` js
class A {
}
class B extends A {
}

function needsAs(as:Array<A>):void {
  as.map(a => {/*...*/});
}

let bs:Array<B> = [new B()];

needsAs(bs);
```

```
8: function needsAs(as:Array<A>):void {
                             ^ A. This type is incompatible with
12: let bs:Array<B> = [new B()];
                 ^ B
```

This can be fixed by changing `needsAs` to `needsAs<T:A>(as:Array<T>):void`. However, it would be more convenient if there was a covariant interface for `Array` such as `ConstArray<+T>` that could be used instead so that the type parameter is not needed.

Similarly for `Map` and `Set`.

**Comment by jesseschalken** [comment]
Ping @avikchaudhuri

**Comment by avikchaudhuri** [comment]
@jesseschalken Completely agree.

**Comment by gnprice** [comment]
There seems to be one of these for Array now, namely `$ReadOnlyArray`.

It's not in the docs anywhere I can find. But it is mentioned in the v0.38.0 release notes: https://github.com/facebook/flow/releases/tag/v0.38.0 and on #2811 .

I don't know a way to do this for Set or Map, though.

**Comment by TrySound** [comment]
[$ReadOnlyMap](https://github.com/facebook/flow/blob/master/lib/core.js#L534)
[$ReadOnlySet](https://github.com/facebook/flow/blob/master/lib/core.js#L534)
[$ReadOnlyArray](https://github.com/facebook/flow/blob/master/lib/core.js#L197)

**Comment by gnprice** [comment]
Thanks @TrySound for that pointer!

Looks like these were added in 1e71f2358 in October, released in v0.86.
